packages/grid/x-data-grid/src/hooks/features/columns/useGridColumns.ts

```diff
@@ -430,24 +433,26 @@ export function useGridColumns(
 
     logger.info(`GridColumns have changed, new length ${props.columns.length}`);
 
+    if (
+      previousColumnsProp.current === props.columns &&
+      previousColumnTypesProp.current === columnTypes
+    ) {
+      return;
```

Comment:
The following test fails in React 18 if we don't check again if the props changed since the last time the state was recreated:

![image](https://user-images.githubusercontent.com/42154031/159138266-26fc66fa-2404-4f15-ae44-d135cf5ad708.png)

The rationale is that the `columns` sets one order for the columns, then `initialState` sets another. We first initialize with `initialState` but the second effect call changes to the `columns`'s order. According to the test, the initial state should win.

Reply:
This seems to be new version of the check just before with `isFirstRender.current` which could be removed

The check has been added in #2782

Reply:
I didn't remove it because it was not breaking anything but yeah it can be removed. All of these `isFirstRender.current` checks won't work with StrictEffects. We can ensure a effect only runs once but we can't ensure it won't be called in the first render. The effects should be resilient to be called more than once: https://github.com/reactwg/react-18/discussions/18